docs/executor_plugins.md

```diff
@@ -0,0 +1,140 @@
+# Executor Plugins
+
+## Types
+
+### Template Executor
+
+There is only one type of executor plugin, one that runs custom ("plugin") templates, e.g. for non-pod tasks such as
+Tekton builds or Spark jobs.
+
+## A Simple Python Plugin
+
+Let's make a Python plugin that prints "hello" each time the workflow is operated on.
+
+We need the following:
+
+1. [Plugins enabled](plugins.md) .
+2. A HTTP server that will be run as a sidecar to the main container and will respond to RPC HTTP requests from the executor.
+3. Configuration so the controller can discover the plugin.
+
+We'll need to create a script that starts a HTTP server:
+
+```python
+import json
+from http.server import BaseHTTPRequestHandler, HTTPServer
+
+
+class Plugin(BaseHTTPRequestHandler):
+
+    def args(self):
+        return json.loads(self.rfile.read(int(self.headers.get('Content-Length'))))
+
+    def reply(self, reply):
+        self.send_response(200)
+        self.end_headers()
+        self.wfile.write(json.dumps(reply).encode("UTF-8"))
+
+    def unsupported(self):
+        self.send_response(404)
+        self.end_headers()
+
+    def do_POST(self):
+        if self.path == '/api/v1/template.execute':
+            args = self.args()
+            if 'hello' in args['template'].get('plugin', {}):
+                self.reply({'node': {'phase': 'Succeeded', 'message': 'Hello template!'}})
+            else:
+                self.reply({})
+        else:
+            self.unsupported()
+
+
+if __name__ == '__main__':
+    httpd = HTTPServer(('', 4355), Plugin)
+    httpd.serve_forever()
+```
+
+Some things to note here:
+
+* You only need to implement the calls you need. Return 404 and it won't be called again.
+* The path is the RPC method name.
+* The request body contains the template's input parameters.
+* The response body contains the node result, including the phase (e.g. "Succeeded" or "Failed") and a message
+
+Next, create a manifest named `plugin.yaml`:
+
+```yaml
+kind: ExecutorPlugin
+metadata:
+  name: hello
+spec:
+  description: This is the "hello world" plugin
+  container:
+    command:
+      - python
+      - -c
+    image: python:alpine3.6
+    name: hello-executor-plugin
+    ports:
+      - containerPort: 4355
+```
+
+Build and install as follows:
```

Comment:
Why do we need argo to build plugins? Why can't users just specify their own images and build them themselves? I guess I don't understand the need for a configmap in the design?

Reply:
I think I'm understanding why... because a plugin will define it's own schema and not depend on a boilerplate schema (such as `template.container` or `template.script`)

Reply:
at the very least "generate" or "codegen" might be a better term than "build"?

Reply:
I chose `build` because `helm build` and `yarn build`. I'm not keen to change it.